File: harness/control_flow/builder.py
from typing import Set
from harness.core import StateGraphNode, StateGraphEdge, Process
from harness.control_flow.node import ControlFlowNode, ControlFlowLabel, ControlFlowBranchNode, ControlFlowStatement, ControlFlowSequence, ControlFlowLabelledNode, ControlFlowGotoNode, ControlFlowSynchronization, ControlFlowInitBarrierNode
from harness.control_flow.mutex import ControlFlowMutexSet

class ControlFlowBuilder:
    def __init__(self, root: StateGraphNode):
        self._root = root
        self._reverse_edges = dict()
        self._backward_edges = set()
        self._labelled_states = dict()
        self._scan()

    @property
    def state_graph_root(self) -> StateGraphNode:
        return self._root
    
    def control_flow(self, process: Process, mutex_set: ControlFlowMutexSet) -> ControlFlowNode:
        node = self._generate_control_flow_node(process, mutex_set, self.state_graph_root, set())
        locked_mutexes = list(mutex_set.locked_in_state(process, self.state_graph_root))
        if locked_mutexes:
            return ControlFlowSequence([
                ControlFlowSynchronization(locked_mutexes, ()),
                ControlFlowInitBarrierNode(),
                node
            ])
        else:
            return node
    
    def _scan(self):
        self._scan_reverse_edges()
        self._scan_backward_edges()

    def _scan_reverse_edges(self):
        for state in self.state_graph_root.all_nodes:
            for edge in state.edges:
                if edge.target not in self._reverse_edges:
                    self._reverse_edges[edge.target] = list()
                self._reverse_edges[edge.target].append(edge)

    def _scan_backward_edges(self):
        visited = set()
        queue = [(self._root, list())]
        while queue:
            current_state, current_path = queue.pop()
            if current_state in visited:
                continue
            visited.add(current_state)
            if current_state not in self._labelled_states:
                self._labelled_states[current_state] = ControlFlowLabel(current_state)
            new_path = [*current_path, current_state]
            for edge in current_state.edges:
                if edge.target in new_path:
                    self._backward_edges.add(edge)
                else:
                    queue.append((edge.target, new_path))

    def _generate_control_flow_node(self, process: Process, mutex_set: ControlFlowMutexSet, state: StateGraphNode, generated_states: Set[StateGraphNode]) -> ControlFlowNode:        
        def generate_edge(edge: StateGraphEdge) -> ControlFlowNode:
            if edge in self._backward_edges or edge.target in generated_states:
                sequence = [
                    ControlFlowStatement(edge),
                    ControlFlowGotoNode(self._labelled_states[edge.target])
                ]
            else:
                sequence = [
                    ControlFlowStatement(edge),
                    self._generate_control_flow_node(process, mutex_set, edge.target, generated_states)
                ]
            
            locked_mutexes = set(mutex_set.locked_in_state(process, edge.target))
            unlocked_mutexes = set(mutex_set.locked_in_state(process, edge.source))

            hold_mutexes = locked_mutexes.intersection(unlocked_mutexes)
            unlocked_mutexes.difference_update(hold_mutexes)
            locked_mutexes.difference_update(hold_mutexes)
            if locked_mutexes or unlocked_mutexes:
                sequence.insert(0, ControlFlowSynchronization(locked_mutexes, unlocked_mutexes, rollback_on_failure=self._labelled_states[edge.source]))
            return ControlFlowSequence(sequence)
        
        generated_states.add(state)

        edge_nodes = list()
        for edge in state.edges:
            edge_nodes.append(generate_edge(edge))

        if len(edge_nodes) > 1:
            node = ControlFlowBranchNode(edge_nodes)
        elif len(edge_nodes) == 1:
            node = edge_nodes[0]
        else:
            node = ControlFlowSequence(())

        node_label = self._labelled_states.get(state, None)
        if node_label is None:
            return node
        else:
            return ControlFlowLabelledNode(
                node_label,
                node
            )

File: harness2/src/harness/control_flow/builder.rs
use std::{collections::{HashMap, HashSet}, rc::Rc};

use crate::harness::core::{error::HarnessError, process::{ProcessID, ProcessSet}, state_machine::{StateMachineContext, StateMachineEdgeID, StateMachineNodeID}};

use super::{mutex::ControlFlowMutexSet, node::{ControlFlowLabel, ControlFlowLabelGenerator, ControlFlowNode}};

pub struct ControlFlowBuilder {
    backward_edges: HashSet<StateMachineEdgeID>,
    labelled_states: HashMap<StateMachineNodeID, ControlFlowLabel>
}

impl ControlFlowBuilder {
    pub fn new(context: &StateMachineContext, root: StateMachineNodeID) -> Result<ControlFlowBuilder, HarnessError> {
        let mut builder = ControlFlowBuilder {
            backward_edges: HashSet::new(),
            labelled_states: HashMap::new()
        };
        builder.initialize(context, root)?;
        Ok(builder)
    }

    pub fn build(&self, context: &StateMachineContext, process_set: &ProcessSet, process: ProcessID, mutex_set: &ControlFlowMutexSet<'_>) -> Result<ControlFlowNode, HarnessError> {
        let process_entry_node = process_set.get_process_entry_node(process)
            .ok_or(HarnessError::new("Unable to find process entry node"))?;
        let locked_initially = mutex_set.get_locked_mutexes_in_state(process, process_entry_node)
            .collect::<HashSet<_>>();
        let node = self.generate_control_flow_node(context, process, process_entry_node, mutex_set, &mut HashSet::new())?;
        if locked_initially.is_empty() {
            Ok(ControlFlowNode::Sequence(Vec::from([
                ControlFlowNode::InitBarrier,
                node
            ])))
        } else {
            Ok(ControlFlowNode::Sequence(Vec::from([
                ControlFlowNode::Synchronization(locked_initially, HashSet::new(), None),
                ControlFlowNode::InitBarrier,
                node
            ])))
        }
    }

    fn initialize(&mut self, context: &StateMachineContext, root: StateMachineNodeID) -> Result<(), HarnessError> {
        let mut label_gen = ControlFlowLabelGenerator::new();
        let mut visited = HashSet::new();
        let mut queue = Vec::from([(root, Rc::new(Vec::new()))]);
        while !queue.is_empty() {
            let (current_state, current_path) = queue.pop().expect("Expected queue to be non-empty");
            if visited.contains(&current_state) {
                continue;
            }
            visited.insert(current_state);

            if !self.labelled_states.contains_key(&current_state) {
                self.labelled_states.insert(current_state, label_gen.next_label());
            }

            let new_path = Rc::new({
                let mut path = current_path.as_ref().clone();
                path.push(current_state);
                path
            });

            let edges = context.get_edges_from(current_state)
                .ok_or(HarnessError::new("Unable to retrieve edges coming from a state"))?;
            for edge in edges {
                let edge_target = context.get_edge_target(edge)
                    .ok_or(HarnessError::new("Unable to retrieve edge target"))?;
                if new_path.contains(&edge_target) {
                    self.backward_edges.insert(edge);
                } else {
                    queue.push((edge_target, new_path.clone()));
                }
            }
        }
        Ok(())
    }

    fn generate_control_flow_node(&self, context: &StateMachineContext, process: ProcessID, state: StateMachineNodeID, mutex_set: &ControlFlowMutexSet<'_>, generated_states: &mut HashSet<StateMachineNodeID>) -> Result<ControlFlowNode, HarnessError> {
        generated_states.insert(state);

        let edges = context.get_edges_from(state)
            .ok_or(HarnessError::new("Unable to retrieve edges originating from a state"))?;
        let mut edge_nodes = edges
            .map(| edge | self.generate_edge(context, process, edge, mutex_set, generated_states))
            .collect::<Result<Vec<_>, HarnessError>>()?;

        let node = if edge_nodes.len() > 1 {
            ControlFlowNode::Branch(edge_nodes)
        } else if !edge_nodes.is_empty() {
            edge_nodes.pop().expect("Expected control-flow node list to be non-empty")
        } else {
            ControlFlowNode::Sequence(Vec::new())
        };

        match self.labelled_states.get(&state) {
            Some(label) =>
                Ok(ControlFlowNode::LabelledNode(*label, node.into())),
            _ => Ok(node)
        }
    }

    fn generate_edge(&self, context: &StateMachineContext, process: ProcessID, edge: StateMachineEdgeID, mutex_set: &ControlFlowMutexSet<'_>, generated_states: &mut HashSet<StateMachineNodeID>) -> Result<ControlFlowNode, HarnessError> {
        let edge_source = context.get_edge_source(edge)
            .ok_or(HarnessError::new("Unable to retrieve edge source"))?;
        let edge_target = context.get_edge_target(edge)
            .ok_or(HarnessError::new("Unable to retrieve edge target"))?;

        let mut sequence = if self.backward_edges.contains(&edge) || generated_states.contains(&edge_target) {
            let label = *self.labelled_states.get(&edge_target).expect("Expected edge target state to have a label");
            Vec::from([
                ControlFlowNode::Statement(edge),
                ControlFlowNode::Goto(label)
            ])
        } else {
            Vec::from([
                ControlFlowNode::Statement(edge),
                self.generate_control_flow_node(context, process, edge_target, mutex_set, generated_states)?
            ])
        };

        let mut locked_mutexes = mutex_set.get_locked_mutexes_in_state(process, edge_target)
            .collect::<HashSet<_>>();
        let mut unlocked_mutexes = mutex_set.get_locked_mutexes_in_state(process, edge_source)
            .collect::<HashSet<_>>();

        let hold_mutexes = locked_mutexes.intersection(&unlocked_mutexes)
            .map(| x | *x)
            .collect::<HashSet<_>>();

        unlocked_mutexes = unlocked_mutexes.difference(&hold_mutexes)
            .map(| x | *x)
            .collect();
        locked_mutexes = locked_mutexes.difference(&hold_mutexes)
            .map(| x | *x)
            .collect();

        if !locked_mutexes.is_empty() || !unlocked_mutexes.is_empty() {
            let label = *self.labelled_states.get(&edge_source).expect("Expected edge target state to have a label");
            sequence.insert(0, ControlFlowNode::Synchronization(locked_mutexes, unlocked_mutexes, Some(label)));
        }
        Ok(ControlFlowNode::Sequence(sequence))
    }
}
